File: 767_Reorganize String.java
// Time complexity: O(n)
// Space complexity:
// O(1) extra space, since we have at most 26 different characters.
// O(n) space for the output string.

class Solution {
    public String reorganizeString(String s) {
        // 1. count letter frequency
        int[] count = new int[26];
        for (int i = 0; i < s.length(); i++) {
            count[s.charAt(i) - 'a'] ++;
        }
        // 2. push into PQ - int[]{count, letter}
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> Integer.compare(b[0], a[0]));
        for (int i = 0; i < count.length; i++) {
            if (count[i] > 0) pq.offer(new int[]{count[i], i});
        }
        // 3. repeatedly pick the most frequent letter
        StringBuilder sb = new StringBuilder();
        int[] prev = null;

        while (!pq.isEmpty() || prev != null) {
            if (pq.isEmpty() && prev != null) return "";

            int[] cur = pq.poll();
            sb.append((char) (cur[1] + 'a'));
            cur[0] --;

            if (prev != null) {
                pq.offer(prev);
                prev = null;
            }
            if (cur[0] > 0) prev = cur;
        }

        return sb.toString();
    }
}

File: 767_Reorganize String.py
# Time Complexity: O(n log k).
# O(n + k log k + n log k), 因为 k <= n，通常可以合并为：O(n log k).

class Solution:
    def reorganizeString(self, s: str) -> str:
        # 1. count - O(n)
        count = Counter(s)

        # 2. init heap (default min-heap in python) - O(k log(k))
        heap = []
        for c, cnt in count.items(): heappush(heap, (-cnt, c))
        
        # 3. Greedy - prioritize the most frequent letter - O(n log(k))
        res, prev, pCnt = [], "", 0

        while heap:
            tCnt, tmp = heappop(heap)

            res.append(tmp)
            tCnt += 1

            if pCnt < 0: heappush(heap, (pCnt, prev))
            prev, pCnt = "", 0

            if tCnt < 0: prev, pCnt = tmp, tCnt

        return "" if len(res) < len(s) else ''.join(res)